File: 227.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        temp = []
        curr_num = 0
        last_sign = '+'
        for i in range(len(s)):
            if s[i].isdigit():
                curr_num  = curr_num*10 + int(s[i])
                if i != len(s)-1: continue
            if s[i] in ['+','-','*','/'] or i == len(s)-1:
                if last_sign == '+':
                    temp.append(curr_num)
                elif last_sign == '-':
                    temp.append(-curr_num)
                elif last_sign == '*':
                    last_num = temp.pop()
                    temp.append(last_num*curr_num)
                elif last_sign == '/':
                    last_num = temp.pop()
                    temp.append(int(last_num/curr_num))
                curr_num = 0
                last_sign = s[i]
        return sum(temp)

File: 227.java
class Solution {
    public int calculate(String s) {
        int res = 0;
        int temp = 0;
        char op = '+';
        Deque<Integer> stack = new LinkedList<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                temp = c-'0';
                while (i+1 < s.length() && Character.isDigit(s.charAt(i+1))) {
                    i++;
                    temp = temp*10+s.charAt(i)-'0';
                }
            }
            if ((!Character.isDigit(c) && c != ' ') || i == s.length()-1) {
                if (op == '+') stack.push(temp);
                else if (op == '-') stack.push(-temp);
                else if (op == '*') stack.push(stack.pop()*temp);
                else if (op == '/') stack.push(stack.pop()/temp);
                temp = 0;
                op = c;
            }
        }
        while (!stack.isEmpty()) {
            res += stack.pop();
        }
        return res;
    }
}